hypothesis-python/src/hypothesis/internal/conjecture/providers.py

```diff
@@ -94,104 +98,119 @@
-        "none",
-        "if",
-        "then",
-        "else",
-        # strings which can be interpreted as a number
-        "0",
-        "1e100",
-        "0..0",
-        "0/0",
-        "1/0",
-        "+0.0",
-        "Infinity",
-        "-Infinity",
-        "Inf",
-        "INF",
-        "NaN",
-        "9" * 30,
-        # common ascii characters
-        ",./;'[]\\-=<>?:\"{}|_+!@#$%^&*()`~",
-        # common unicode characters
-        "Ω≈ç√∫˜µ≤≥÷åß∂ƒ©˙∆˚¬…æœ∑´®†¥¨ˆøπ“‘¡™£¢∞§¶•ªº–≠¸˛Ç◊ı˜Â¯˘¿ÅÍÎÏ˝ÓÔÒÚÆ☃Œ„´‰ˇÁ¨ˆØ∏”’`⁄€‹›ﬁﬂ‡°·‚—±",
-        # characters which increase in length when lowercased
-        "Ⱥ",
-        "Ⱦ",
-        # ligatures
-        "æœÆŒﬀʤʨß"
-        # emoticons
-        "(╯°□°）╯︵ ┻━┻)",
-        # emojis
-        "😍",
-        "🇺🇸",
-        # emoji modifiers
-        "🏻"  # U+1F3FB Light Skin Tone,
-        "👍🏻",  # 👍 followed by U+1F3FB
-        # RTL text
-        "الكل في المجمو عة",
-        # Ogham text, which contains the only character in the Space Separators
-        # unicode category (Zs) that isn't visually blank:  .  # noqa: RUF003
-        "᚛ᚄᚓᚐᚋᚒᚄ ᚑᚄᚂᚑᚏᚅ᚜",
-        # readable variations on text (bolt/italic/script)
-        "𝐓𝐡𝐞 𝐪𝐮𝐢𝐜𝐤 𝐛𝐫𝐨𝐰𝐧 𝐟𝐨𝐱 𝐣𝐮𝐦𝐩𝐬 𝐨𝐯𝐞𝐫 𝐭𝐡𝐞 𝐥𝐚𝐳𝐲 𝐝𝐨𝐠",
-        "𝕿𝖍𝖊 𝖖𝖚𝖎𝖈𝖐 𝖇𝖗𝖔𝖜𝖓 𝖋𝖔𝖝 𝖏𝖚𝖒𝖕𝖘 𝖔𝖛𝖊𝖗 𝖙𝖍𝖊 𝖑𝖆𝖟𝖞 𝖉𝖔𝖌",
-        "𝑻𝒉𝒆 𝒒𝒖𝒊𝒄𝒌 𝒃𝒓𝒐𝒘𝒏 𝒇𝒐𝒙 𝒋𝒖𝒎𝒑𝒔 𝒐𝒗𝒆𝒓 𝒕𝒉𝒆 𝒍𝒂𝒛𝒚 𝒅𝒐𝒈",
-        "𝓣𝓱𝓮 𝓺𝓾𝓲𝓬𝓴 𝓫𝓻𝓸𝔀𝓷 𝓯𝓸𝔁 𝓳𝓾𝓶𝓹𝓼 𝓸𝓿𝓮𝓻 𝓽𝓱𝓮 𝓵𝓪𝔃𝔂 𝓭𝓸𝓰",
-        "𝕋𝕙𝕖 𝕢𝕦𝕚𝕔𝕜 𝕓𝕣𝕠𝕨𝕟 𝕗𝕠𝕩 𝕛𝕦𝕞𝕡𝕤 𝕠𝕧𝕖𝕣 𝕥𝕙𝕖 𝕝𝕒𝕫𝕪 𝕕𝕠𝕘",
-        # upsidown text
-        "ʇǝɯɐ ʇᴉs ɹolop ɯnsdᴉ ɯǝɹo˥",
-        # reserved strings in windows
-        "NUL",
-        "COM1",
-        "LPT1",
-        # scunthorpe problem
-        "Scunthorpe",
-        # zalgo text
-        "Ṱ̺̺̕o͞ ̷i̲̬͇̪͙n̝̗͕v̟̜̘̦͟o̶̙̰̠kè͚̮̺̪̹̱̤ ̖t̝͕̳̣̻̪͞h̼͓̲̦̳̘̲e͇̣̰̦̬͎ ̢̼̻̱̘h͚͎͙̜̣̲ͅi̦̲̣̰̤v̻͍e̺̭̳̪̰-m̢iͅn̖̺̞̲̯̰d̵̼̟͙̩̼̘̳ ̞̥̱̳̭r̛̗̘e͙p͠r̼̞̻̭̗e̺̠̣͟s̘͇̳͍̝͉e͉̥̯̞̲͚̬͜ǹ̬͎͎̟̖͇̤t͍̬̤͓̼̭͘ͅi̪̱n͠g̴͉ ͏͉ͅc̬̟h͡a̫̻̯͘o̫̟̖͍̙̝͉s̗̦̲.̨̹͈̣",
-        #
-        # examples from https://faultlore.com/blah/text-hates-you/
-        "मनीष منش",
-        "पन्ह पन्ह त्र र्च कृकृ ड्ड न्हृे إلا بسم الله",
-        "lorem لا بسم الله ipsum 你好1234你好",
-    ],
-    key=len,
-)
 
-# Masks for masking off the first byte of an n-bit buffer.
-# The appropriate mask is stored at position n % 8.
-BYTE_MASKS = [(1 << n) - 1 for n in range(8)]
-BYTE_MASKS[0] = 255
+GLOBAL_CONSTANTS: "ConstantsT" = {
+    "float": {float_to_int(f) for f in _constant_floats},
+    "string": set(_constant_strings),
+    "integer": set(),
+    "bytes": set(),
+}
+
+
+_local_constants_hash: Optional[int] = None
+
+
+def _get_local_constants():
+    global _local_constants_hash
+
+    constants = local_constants()
+    constants_hash = hash(tuple((k, tuple(v)) for k, v in constants.items()))
+    # if we've added new constants since the last time we checked, invalidate
+    # the cache.
+    if constants_hash != _local_constants_hash:
+        CONSTANTS_CACHE.cache.clear()
+        _local_constants_hash = constants_hash
+
+    return constants
```

Comment:
I'm not seeing how this caching helps; it looks like we never read from it?

Reply:
`CONSTANTS_CACHE` is written to and read by `_maybe_draw_constant`? It's caching the `choice_permitted` constants for a particular set of constraints.